openedx/core/djangoapps/content/learning_sequences/admin.py

```diff
@@ -229,15 +229,15 @@ def _publish_report_attr(self, course_context, attr_name):
 
     def num_errors(self, course_context):
         return self._publish_report_attr(course_context, 'num_errors')
-    num_errors.short_description = "Errors"
+    setattr(num_errors, "short_description", "Errors")
 
     def num_sections(self, course_context):
         return self._publish_report_attr(course_context, 'num_sections')
-    num_sections.short_description = "Sections"
+    setattr(num_sections, "short_description", "Sections")
 
     def num_sequences(self, course_context):
         return self._publish_report_attr(course_context, 'num_sequences')
-    num_sequences.short_description = "Sequences"
+    setattr(num_sequences, "short_description", "Sequences")
```

Comment:
Well, mypy runs here on the entire `learning_sequences/` directory, including this admin.py module. Without these changes, mypy reports the following errors:

```
openedx/core/djangoapps/content/learning_sequences/admin.py:87: error: "Callable[[CourseSectionSequenceInline, Any], Any]" has no attribute "short_description"
```

Reply:
@ormsbee When you run mypy on a module or directory, it's not just checking that all the annotations are correct--it's stricter than that. mypy is checking that your code can be deduced as properly typed given (1) what it can infer about your code on its own and (2) what it is being told by that annotations that you add. Code without any annotations will generally fail  type checking, since mypy can only infer so much without help.

Specifically, these attributes-assigned-on-functions seem like one of those instances where the dynamicism of Python and Django butts up against the formality of type checking. I assume @regisb made this change because mypy was saying something to the effect of *"I have no reason to believe that `num_sequences`, which is a function, should have an attribute `short_description`."* Using `setattr` tends to fix errors like that, since mypy throws its hands up in the air when it sees dynamic attribute assignment/retrieval and just assumes you know what you're doing.

If we don't like the setattr solution, a couple alternative fixes could be throwing `# type: ignore` on these (I think that would work, but I haven't tried) or [using a type decorator](https://github.com/python/mypy/issues/2087#issuecomment-587741762), as others in this same situation have. I encourage us to think carefully about this, since whatever we go with, we're setting a pattern for every `admin.py` in the repository to follow :)

Reply:
Merde. `setattr(num_sequences, "key", "value")` triggers a pylint warning... https://build.testeng.edx.org/job/edx-platform-quality-pipeline-pr/28019/pylint/new/folder.1177700281/source.4b0778cb-34d0-42ff-8a47-99b2d29d2194/#87

Reply:
FWIW, I have no strong feelings on ignoring pylint vs. ignoring mypy for this section of code, and I'd be fine with rearranging so all of these are grouped at the end of the class definition if it makes it easier.

Reply:
@regisb could you either follow the [ActionWithAttributes example](https://github.com/python/mypy/issues/2087#issuecomment-587741762) (could go in openedx/core/types.py so that other apps could use it) or just mypy-ignore these lines?

The `setattr` approach feels to me like an instance of making the code clunkier in order to make the tooling happy.

Reply:
I went with another approach, based on `Protocol`, referenced in this same thread. Let me know what you think.

Reply:
@regisb Your approach looks great. Nice and straightforward, but also enforces the types of the method attributes, which is cool. I think the class and decorator should live in `openedx/core/types.py` so that other apps' admin pages can use it; do you agree?

Fwiw, Django 3.2 will introduce a new way of customizing admin columns:
```python
 @admin.display(short_description="Record Created")
 def created(self):
        return ...
```
which probably won't bother mypy. So, we could drop your workaround in the medium-term future if we wanted to.

Reply:
This is good to know. Thus I decided to replace `admin_method` with a future-proof `admin_display` decorator. I moved it to a `admin.py` file in the `types` package.